daml/RealFac/Contract/Contract.daml

```diff
@@ -0,0 +1,162 @@
+module RealFac.Contract.Contract where
+
+import DA.Action
+import DA.List
+import RealFac.CashAccount.CashAccount
+import RealFac.Contract.ContractParties
+import RealFac.Contract.Phase
+
+template Contract
+  with
+    contractParties : ContractParties
+    ref : Text
+    phases : [Phase]
+    -- derived tacking data
+    done : Bool
+    toSettle : Int
+  where
+    signatory parties contractParties
+    observer parties contractParties
+    key (contractParties.accountant, ref) : (Party, Text)
+    maintainer key._1
+
+    let
+      ContractParties accountant buyer developer = contractParties
+
+    controller buyer can
+
+      Ack
+        : ContractId Contract
+        with
+          phaseId : Text
+        do
+          let
+            (newPhases, phase, allDone) = processPhase phaseId Done phases
+          create this with
+            phases = newPhases
+            done = allDone
+            toSettle = toSettle + phase.descriptor.cashOnComplete
+
+      Reject
+        : ContractId Contract
+        with
+          phaseId : Text
+        do
+          let
+            (newPhases, phase, allDone) = processPhase phaseId InProgress phases
+          create this with
+            phases = newPhases
+
+    controller developer can
+
+      Start
+        : ContractId Contract
+        with
+          phaseId : Text
+        do
+          let
+            (newPhases, phase, allDone) = processPhase phaseId InProgress phases
+          create this with
+            phases = newPhases
+
+      Complete
+        : ContractId Contract
+        with
+          phaseId : Text
+        do
+          let
+            (newPhases, phase, allDone) = processPhase phaseId PendingDone phases
+          create this with
+            phases = newPhases
+
+    controller accountant can
+
+      nonconsuming Settle
+        : ()
+        do
+          when (toSettle > 0) do
+            exerciseByKey @CashAccount (accountant, buyer) 
+              WithdrawReserved with amount = toSettle
+            exerciseByKey @CashAccount (accountant, developer) 
+              Deposit with amount = toSettle
+            archive self
+            create this with
```

Comment:
This is just gut feel, I would need to play with the model a bit more to get an idea so take this with a grain of salt: It feels like `Contract` has quite a few different states. IME, you are often better off, trying to represent different states with different templates, e.g., one where is something to settle, one for the state where everything is settled (or just archive it if you can), …. Not quite sure how the split looks like here but perhaps something to think about.

Reply:
agreed, covered this in my other comment